### PR TITLE
Adds support for vanilla Gerrit

### LINKED ISSUE #1787: Remove '-review' from Gerrit Remote reviewDomain()
Please remove the addition of '-review' to the subdomain portion of the get reviewDomain() function.

It is potentially breaking for Gerrit installations that are at `gerrit.example.com` instead of `gerrit-review.example.com`

This should either be a configurable option or not assumed at all.

### LINKED ISSUE #1902: Support replacing mirror/replica domain with main domain for remote provider
In our setup with Gerrit (however, I believe this may apply to other types as well) there is one main domain and many local replicas. The replicas are used for fetching, the main one is used for pushing and other Gerrit operations (including REST API, review query, etc.)

It seems however GitLens doesn't support such an approach. Instead, it keeps using the fetch remote to form the `Gerrit Id:` links. Moreover, due to issue #1787 even the mirror domain is broken.

### LINKED ISSUE #1953: Support vanilla Gerrit (which is different from Google Source)
The current Gerrit provider implementation is very tied to Google Source. Google Source is an implementation on top of Gerrit, but many other orgs implement just vanilla Gerrit.

A public vanilla Gerrit to be used as example is the [GerritHub](https://gerrithub.io).

PS: PR incoming.

### PR DESCRIPTION
**Pull request by felipecrs**
# Description

I work with Gerrit every day, and today I realized how much time it could save if Gitlens worked for me. To my surprise, I found Gerrit in the list of supported providers, but when trying to use it, I stumbled upon some issues.

The current Gerrit provider implementation is very tied to Google Source, but Google Source is an implementation on top of the vanilla Gerrit, and many other orgs implements Gerrit in the vanilla way.

This PR turns the existing Gerrit provider into a generic/vanilla Gerrit one (not tied to Google Source), but adds back Google Source as its own provider, meant to be used in Android, Chromium, Gerrit itself, Golang, and other projects from them.

Also, the new provider comes with a new pre-configured domain, which is the [GerritHub](https://gerrithub.io), a publicly known and freely available Gerrit instance.

Where I work, however, we host our own Gerrit instance, and I would configure Gitlens as follows:

```jsonc
"gitlens.remotes": [{ "domain": "gerrit.my-org.com", "type": "Gerrit" }]
```

Closes #1953
Closes #1787
Closes #1788
Closes #1902

# Checklist

- [x] I have followed the guidelines in the Contributing document
- [x] My changes follow the coding style of this project
- [x] My changes build without any errors or warnings
- [x] My changes have been formatted and linted
- [x] My changes include any required corresponding changes to the documentation
- [x] My changes have been rebased and squashed to the minimal number (typically 1) of relevant commits
- [x] My changes have a descriptive commit message with a short title, including a `Fixes $XXX -` or `Closes #XXX -` prefix to auto-close the issue that your PR addresses

### CODE REVIEW

**Comment by felipecrs** [comment]
This is ready to review.

**Comment by andrewsavage1** [comment]
No issues with this from me, thanks for the fix @felipecrs.

**Comment by felipecrs** [comment]
@Levistator, @adambadura as you both reported similar issues, I would kindly ask if this would solve your issues.

**Comment by felipecrs** [comment]
If you would like to test, I built a vsix that you can install in your VS Code:

[gitlens-12.0.5.vsix.zip](https://github.com/gitkraken/vscode-gitlens/files/8447487/gitlens-12.0.5.vsix.zip)

It's not a zip file, simply rename to remove the .zip suffix after downloading (GitHub only allows to upload certain extensions in comments).

File: src/git/remotes/google-source.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+	override get name() {
+		return this.formatName('Google Source');
+	}
+
+	protected override get baseUrl(): string {
+		return `${this.protocol}://${this.domain}/${this.path}`;
+	}
+
+	private get reviewDomain(): string {
+		const [subdomain, secondLevelDomain, topLevelDomain] = this.domain.split('.');
```
Comment (acarlson1029) [inline src/git/remotes/google-source.ts:21]: FWIW this doesn't work for domains with more than three levels.

e.g. `project.git.source.company-tld.com`

Can this just split and mutate the subdomain?

Comment (felipecrs) [reply src/git/remotes/google-source.ts:21]: This is only for GoogleSource. Are you sure this is applicable for you?

**Comment by adambadura** [comment]
> If you would like to test, I built a vsix that you can install in your VS Code:

Is there any document describing how to use your package? Especially in remote scenarios? I'm sorry, but I'm new to Visual Studio Code.

**Comment by adambadura** [comment]
> > If you would like to test, I built a vsix that you can install in your VS Code:
> 
> Is there any document describing how to use your package? Especially in remote scenarios? I'm sorry, but I'm new to Visual Studio Code.

Turns out on the "Extensions" tab, in the "..." menu there is "Install from VSIX..." and it works as expected, including my remote work scenario.

**Comment by adambadura** [comment]
> @Levistator, @adambadura as you both reported similar issues, I would kindly ask if this would solve your issues.

The issue #1787 (on adding `-review` in domain) seems to indeed be closed.

However, my own issue #1902 is not closed. Now I have `gerrit-wrsl1.int.net.nokia.com` used as domain. This matches the remote of Git. However, links to changes etc. should (in our setup) be done for the main server (`gerrit.ext.net.nokia.com`) instead. Only fetching is done from the mirror server.

Or do I have to do some extra configuration in the plugin settings to have this covered?

Interestingly, this could perhaps be deduced (or wild-guessed) from Git configuration itself. My `.gitconfig` contains entries like:
```
[url "https://<USER>@gerrit-wrsl1.int.net.nokia.com:29418"]
        insteadOf = https://<USER>@gerrit.ext.net.nokia.com:29418
[url "https://<USER>@gerrit.ext.net.nokia.com:29418"]
        pushInsteadOf = https://<USER>@gerrit.ext.net.nokia.com:29418
        pushInsteadOf = https://<USER>@gerrit-wrsl1.int.net.nokia.com:29418
```
including also versions with `ssh://` and without the `<USER>` (although I'm not sure if I indeed need all 4 variants...).

**Comment by adambadura** [comment]
For the record, my `C:/Users/(...)/AppData/Roaming/Code/User/settings.json` has the following:
```json
    "gitlens.remotes": [
        {"domain": "gerrit.ext.net.nokia.com", "type": "Gerrit"},
        {"domain": "gerrit-wrsl1.int.net.nokia.com", "type": "Gerrit"}
    ],
```

**Comment by felipecrs** [comment]
@adambadura Would you mind running:

```console
git remote -v
```

From within your cloned project?

**Comment by felipecrs** [comment]
Where I work, we have a comparable situation. But there is one thing weird in your case, the protocol is set as HTTPS, but the port 29148 (which is the Gerrit's known SSH port).

If my suspicious is correct, you are using SSH in the `git remote`. Internally, we also need to fetch from mirror and push to master, but only when using SSH. For HTTPS, the load balancing from Gerrit side happens transparently.

I have switched from SSH to HTTPS since a long time ago and I never looked back. With GCM, I don't need to re-type my password either.

Perhaps it can work for you too, although I don't deny we can improve the extension to support such a mirror/master scenario.

**Comment by adambadura** [comment]
> ```
> git remote -v
> ```

Sure:
```
origin  ssh://<USER>@gerrit-wrsl1.int.net.nokia.com:29418/MN/5G/NB/gnb (fetch)
origin  ssh://<USER>@gerrit.ext.net.nokia.com:29418/MN/5G/NB/gnb (push)
```

**Comment by adambadura** [comment]
> ```
> git remote -v
> ```

Sure:
```
origin  ssh://<USER>@gerrit-wrsl1.int.net.nokia.com:29418/MN/5G/NB/gnb (fetch)
origin  ssh://abadura@gerrit.ext.net.nokia.com:29418/MN/5G/NB/gnb (push)
```

**Comment by felipecrs** [comment]
@adambadura, I tested here, locally, and it seems to be working as expected. There is one point, however:

```jsonc
"gitlens.remotes": [
        {"domain": "gerrit.ext.net.nokia.com", "type": "Gerrit"},
    ],
```

Do **NOT** add the mirror to the list of Gitlens Remotes. Can you please check if that works for you?

**Comment by adambadura** [comment]
> Can you please check if that works for you?

I will get back to you on this on Monday.

File: src/git/remotes/google-source.ts (hunk 1 above)
Comment (acarlson1029) [reply src/git/remotes/google-source.ts:21]: Yes, it was serendipitous that I found this commit because I was just debugging this issue at the time.

Comment (acarlson1029) [reply src/git/remotes/google-source.ts:21]: Upon double-checking, I guess it's not strictly necessary.
The domain I was using also resolves with `project.googlesource.com`

Comment (felipecrs) [reply src/git/remotes/google-source.ts:21]: > The domain I was using also resolves with project.googlesource.com

Not sure if I understood this part.

Anyway, after this PR, you would configure VS Code with:

```json
"gitlens.remotes": [{ "domain": "project.git.source.company-tld.com", "type": "Gerrit" }]
```

If you have such an environment to test, I would advise you to test it before we merge. The .vsix is available here:

https://github.com/gitkraken/vscode-gitlens/pull/1954#issuecomment-1092309335

**Comment by Levistator** [comment]
Yep it works for me. Thanks @felipecrs !!

**Comment by adambadura** [comment]
> > Can you please check if that works for you?
> 
> I will get back to you on this on Monday.

Indeed, using only the main server:
```json
    "gitlens.remotes": [
        {"domain": "gerrit.ext.net.nokia.com", "type": "Gerrit"}
     ],
```
helped in that the proper server is reached (at least from the `Change-Id` link).

However, as it turned out, this is not enough in my case. The server is OK, but the web access is done through `gerrit.ext.net.nokia.com/gerrit/` instead. So, the `Change-Id` link should be
```plain
https://gerrit.ext.net.nokia.com/gerrit/q/<hash>
```
while the plugin makes it
```plain
https://gerrit.ext.net.nokia.com/q/<hash>
```

I know this goes beyond what I described in #1902 and I would close #1902 with this one and open a more specific issue afterward. However, perhaps you do know already how to address this configuration?

I did try using `"gerrit.ext.net.nokia.com/gerrit"` for the `"domain"`, but unsurprisingly it didn't work. Moreover, the entire `Change-Id` linking functionality ceased to work (no link made at all), so I expect this must be some configuration error.

**Comment by felipecrs** [comment]
@adambadura what would be the Gitiles homepage for your example then?

https://gerrit.ext.net.nokia.com/gerrit/plugins/gitiles

Is it the above?

**Comment by adambadura** [comment]
> @adambadura what would be the Gitiles homepage for your server then?
> 
> 1. https://gerrit.ext.net.nokia.com/
> 2. https://gerrit.ext.net.nokia.com/plugins/gitiles
> 3. https://gerrit.ext.net.nokia.com/gerrit/plugins/gitiles
> 
> Is it any of the above?

It is the last one, although I had to add the trailing / as otherwise I was getting "Not Found" error.

**Comment by adambadura** [comment]
OK then, let's close #1902 along with this one.

**Comment by felipecrs** [comment]
@eamodio just to say this is ready to review now.

**Comment by eamodio** [comment]
Wow, @felipecrs, this is awesome. Thank you so much for your contribution! And thank you to all the testers too!

Thank you!

**Pull request merged**